### ISSUE TITLE
No option for enabling/disabling mods in games.

### ISSUE THREAD

**Issue by Rencer** [minetest/minetest#8622]
I'm playing Minetest with mods in Manjaro Linux.
I installed Minetest from the AUR repo and downloaded some games and mods from the ContentDB site to try them out.

I can't figure out how to enable/disable mods in the downloaded games. When I click on the name of the mod the button to enable/disable them aren't there at the top of the screen, as it should be.
So how I can enable or disable a mod in a game?

When I install mods manually they all can be enabled/disabled in the configure menu, only in those downloaded games that can't let me enable/disable the individual mods that are in the game.

**Comment by DS-Minetest** [comment]
Subgame mods aren't thought to be enabled or disabled.

**Comment by benrob0329** [comment]
All the mods are intended to run in the game. It's not usually a collection of optional modules, but rather necessary content to make the game whole.

**Comment by Rencer** [comment]
Thanks, I didn't know that. I thought, I can enable/disable individual mods.
I figured it out I can remove mods if I simply delete them from the game's mod folder, or copy them elsewhere so I can add it back if they needed. But having an option to enable/disable them from the game menu makes that easier.
Any particular reason why we can't enable/disable mods if they are come with a modpack? It is a failsafe mechanic so we can't broke the modpacks...I mean games. Or somthing technical?

**Comment by Rencer** [comment]
> All the mods are intended to run in the game. It's not usually a collection of optional modules, but rather necessary content to make the game whole.

Oh, so just as I thought. Thanks, you just answered my question before I even posted it!

**Comment by pauloue** [comment]
I think this is a valid feature request. Being able to disable game mods would be quite useful.

**Comment by DS-Minetest** [comment]
Replacing mods with mods of the same name might be useful.

@Rencer There are also modpacks. Disabling individual mods of a modpack is possible.

**Comment by Rencer** [comment]
> Replacing mods with mods of the same name might be useful.
> 
> @Rencer There are also modpacks. Disabling individual mods of a modpack is possible.

Modpacks?
Isn't those "games" are the modpacks and just called "games" in Minetest?

From the official Wiki:
"Internally, a game itself a collection of (pre-packaged) mods and a few extra files (mostly for the main menu images), but packaged together to provide a complete interactive experience."
So, if I understand it correctly, those "games" are the "modpacks", or not?
Now I'm totally confused...

If I'm correct and what I downloaded, that "game" is a "modpack" than why I can't enable/disable the individual mods in that game? I only can enable/disable individual mods that I downloaded and are in the /mods folder.

**Comment by style-nine** [comment]
All the games I've ever put together would certainly break if you enabled/disabled mods willy-nilly. That said, it might be a cool feature to have.

**Comment by DS-Minetest** [comment]
Subgames are not modpacks.
You always have one subgame (or "game") per world (eg. `minetest_game` or `realtest`). Then you can also activate additional mods for this world (eg. `moreores`). Some mods (eg. `mesecons_pistons`) are in a modpack (in this example the mesecons modpack). You can activate a modpack in a whole in the world configuration, but you don't have to.

**Comment by pauloue** [comment]
@style-nine This would have to be implemented in such a way that disabling a mod automatically disables all mods that depend on it. For game mods which must be loaded always (e.g. `default`), `mod.conf` could support something like `required = true` to disallow disabling.

**Comment by Rencer** [comment]
Now I get it, what is the difference between game and modpack.
So that we cannot disable/enable mods in a subgame is a failsafe, so we wont broke the whole game. It is a good thing. But there are plenty of mods that are just astetics, or somthing that a person does not like.

In my case, in the SkyFactory game it is the WiTT mod, that is similar to Waila in Minecraft.
I can't even tell you guys how much I hate that mod...That is that mod that literally made me to never-ever watch any letsplay videos ever again. I just can't stand that when you walk around, or just turn around and useless texts constantly flashing on the screen. I simply can't understand that (and any similar) mod why anyone want to use it. It's useless because it doesn't help anything if I see what is the name of that block and what mod add it to the game. If I don't know how to use that block or what I can do with it, knowing the name that block and what mod it came from doesn't help in any way.
I already deleted WiTT from the SkyFactory game, and it still works perfectly.
As you can see, not all the mods are required, so it was be a really cool feature if we can enable/disable mods in games.

Thanks you guys to change this to a feature request.

**Comment by orwell96** [comment]
> In my case, in the SkyFactory game it is the WiTT mod, that is similar to Waila in Minecraft.
Actually, I don't quite get why WiTT is part of a subgame. You should suggest to the subgame developer to remove the mod.

A game is a collection of mods and modpacks. It provides a basic set of objects and features to provide a game experience. Individual mods/modpacks in a game are not meant to be disabled because they are part of the playing experience. However, it also should not add additional utilities which are not required by the game concept (such as WiTT mentioned). A game itself is not a modpack.

A modpack is a collection of mods that somehow belong together, but do not provide a basic playable game. Essentially it is just a directory containing mods. You can freely disable individual mods in a modpack.

Regarding disabling game mods: I would rather like a "list of recommended mods" that a game author recommends for his game, but which are not mandatory. They would be installed just as any regular mod. The game would then only contain "mandatory" mods which do not make sense to be disabled.

**Comment by paramat** [comment]
> Individual mods/modpacks in a game are not meant to be disabled because they are part of the playing experience.

There's no reason they can't be disableable, as a way of customising the game.
For example, in MTGame we are discussing adding mobs. Not all players want them so the simplest way to decide is disable the mod, this avoids having to add code and settings.
Very useful for developers too.
I support this +1

**Comment by LoneWolfHT** [comment]
This will be useful for when the craftguide mod is added too

**Comment by benrob0329** [comment]
>There's no reason they can't be disableable, as a way of customising the game.

If the game breaks because you disabled needed content, users will complain. It's better to add settings options for any _actually_ optional content, which may also enable other things in the game to fill the needed gap.

**Comment by Rencer** [comment]
> > There's no reason they can't be disableable, as a way of customising the game.
> 
> If the game breaks because you disabled needed content, users will complain. It's better to add settings options for any _actually_ optional content, which may also enable other things in the game to fill the needed gap.

I think, what pauloue suggested 2 days ago is a good way to solve this thing. I think it isn't that hard to implement a method that somehow register which mod is a must have and which can be disabled without breaking the game.
In the Minecraft FTB launcer this thing is already implemented. There are some Minecraft modpacks for the FTB launcher that contains several mods that are disabled by default. In that launcher there is a option for edit the modpack, and there you can see what mods are enabled and what are disabled by default. Those that are disabled by default are some extra mods, or variants of some mod that do the same thing, so you can add some extra to the pack or switch to another mod that you prefer more than the default one.
Plus, there is an option for manually adding even more mods that aren't part of that pack.
I really like to see something similar in Minetest too.
Only problem with that, that you can freely enable/disable everything there and if you don't exactly know what you doing, you can break the modpack.
It is obvious that mods,modpacks that are required for the subgame to work correctly can't be disabled. There must be a way to prevent this, something like pauloue mentioned.

**Comment by benrob0329** [comment]
**Games are not modpacks.** Modpacks are a collection of separate mods which distributed together so that the player doesn't have to install them separately. Games are a single piece of base content meant to be played as a whole, with some potential settings for tweaking gameplay.

People already don't see much of a different between a game and a modpack, this will only make that worse.

**Comment by paramat** [comment]
> I think it isn't that hard to implement a method that somehow register which mod is a must have and which can be disabled without breaking the game.

That may be necessary.

**Comment by Rencer** [comment]
> **Games are not modpacks.** Modpacks are a collection of separate mods which distributed together so that the player doesn't have to install them separately. Games are a single piece of base content meant to be played as a whole, with some potential settings for tweaking gameplay.
> 
> People already don't see much of a different between a game and a modpack, this will only make that worse.

We all already know that!!! Why you keep repeating this? 
Post something useful, please....

**Comment by benrob0329** [comment]
Ok, here's a real world example. The TNT mod from MTG shouldn't be completely disabled because some mods need the functions from it. If disabling per-game mods becomes the standard way of disabling things in game, then it will be inconsistent with current and necessary behavior. It allows players to easily break the game from inside the menu, with no good warning about it.

Yes, adding a `required` mod.conf settings is handy, but what for mods which _are_ optional, but shoudn't be disabled completely for the functions they define? If it's required, users will be confused because optional content can be disabled, if it's not, users may break the game. By doing this you split the game-specific customization options between the settings and mod menus, making the act of actually customizing the game more complicated.

What would be better would be to give each world their own conf, and then display the game-defined options in the customize menu for easy access.

**Comment by paramat** [comment]
Yes good points, it's actually difficult to decide if a mod in a game should be force-enabled or not, because this depends on the situation.
So, if we added this feature, i would actually prefer no safeguards, 'disable at your own risk'.
I'm more unsure now so removing my support, but not opposed.

**Comment by Rencer** [comment]
> Ok, here's a real world example. The TNT mod from MTG shouldn't be completely disabled because some mods need the functions from it. If disabling per-game mods becomes the standard way of disabling things in game, then it will be inconsistent with current and necessary behavior. It allows players to easily break the game from inside the menu, with no good warning about it.
> 
> Yes, adding a `required` mod.conf settings is handy, but what for mods which _are_ optional, but shouldn't be disabled completely for the functions they define? If it's required, users will be confused because optional content can normally be disabled, if it's not, users may break the game. By doing this you split the game-specific customization options between the settings and mod menus, making the act of actually customizing the game more complicated.
> 
> What would be better would be to give each world their own conf, and then display the game-defined options in the customize menu for easy access.

I don't think that the ability to disable an individual mod is that so "dangerous" and definitely don't have to be so complicated as you said.
If any individual mod or something required by another mod to function correctly, than, simply, that mod is NOT an optional mod, but required, so you can't disable it. That simple.
This `required` mod.conf setting must be set by the maker(s) of the given subgame, because they are who knows exactly what mod is a must-have and which ones are that are really just optional mods.
This config setting must be a list that contains everything that is a must have.
In this way all the mods that are really required cannot be disabled.
What mod is a must have and which just really optional can vary by subgame.

Only one thing that I can't understand in your example.
What do you mean by "split the game-specific customization options between the settings and mod menus"?
Maybe I miss something, but I have only one place where I can enable/disable mods; when I selected a subgame in the bottom of the main menu, and then click on a word that is already created, and then click on the configuration button. Only there, where I can eanble/disable mods.
Can you explain why you think the ability to enable/disable mods ar split the customization?
If this requirements are set in a config file, and that config file come with the game, I don't see any problem that can break anything. It just add the functionality in the subgame's configuration menu to be enable/disable those mods that aren't listed in the mod.conf that includes all the must-have mods.

**Comment by benrob0329** [comment]
Games can define custom settings that will be listed at the very bottom of the "All Settings" menu. The TNT mod from MTG can be disabled this way, however the functions it defines are used by other mods (even if the tnt block isn't registered), so the mod shouldn't be disabled completely. By making the standard way of disabling in-game content to be to disable mods, it will only make things more confusing when there is a reason to use settings to do so.

This is why I also suggested displaying these settings inside the customize menu, which I think would help alleviate the problem this PR was created for.

**Comment by LoneWolfHT** [comment]
Why not a game.conf setting similar to
```
disableable_mods = craftguide, tnt
```

**Comment by paramat** [comment]
I agree that there is a danger to using the disabling of mods as a way to implement options. Settings will be the best way in many cases.
And to repeat, it's impossible to know what mod is required or not, so i oppose any kind of listing of disableable mods to avoid ineffective bloat.
I would only support a very simple disabling at the users own risk.

**Comment by benrob0329** [comment]
@LoneWolfHT why not just allow game-specific options in world.mt, rather than outright disabling entire chunks of code forcefully?

**Comment by LoneWolfHT** [comment]
My idea was to just allow mods to be disabled via the configure tab if they were listed in this line?
`disableable_mods = craftguide, tnt`

**Comment by benrob0329** [comment]
Again though, the `tnt` mod shouldn't be completely disabled because it defines functions which other mods use. The same could be true for any other mod, which is why settings are best for customization.

**Comment by Rencer** [comment]
When you look in to any mods folder there is a file (at least the few games that I already tried, all have this file) called `depends.txt`
When you look into that file, you see other mods listed there. Isn't that file is there exactly for the reason to check if a mod required another mod to function?
That file can be used to check if a mod required by other mods (depends on it) so we can know which mod cannot be disabled. And as I suggested, subgame makers have to provide some-kind of list that contains all the key mods, modpacks that are a must-have, so they cannot be disabled.
I think we only need is that list that the subgame author provide and contains the mods that cannot be disabled.
So, if subgames come with a list that contains all the mods that cannot be disabled, combining that with the already existing `depends.txt`, we can determine which mods really cannot be disabled.

If some is so stubborn and can't understand that; if you cannot disable a mod normally (when you already selected the subgame and select the word you want to play, then click on the configure button) but go into advanced settings and forcibly disable a must-have mod...Well, than the problem is on the user side, there must be something seriously wrong in that persons head...
Come on, it's is even in the name; advanced settings. That already indicates that you have to be very careful, and you must need an advanced knowledge about what you can do with those settings.
So all the responsibility is on the user who mess with the advanced settings, so if the user broke something that is only her/his problem.
As far as I know everywhere in this world, both RL and even in softwares, if you broke something because you mess with a thing that you can't understand, you are on your own. You loose the warranty...
I understand that you guys like to have a completely fail-safe system, but it is impossible to do.
Especially when we already have this advanced settings place, what everyone can access without restrictions and can modify all the settings.
I think most of those who plays Minetest are somewhat more advanced users, than just a plain everyday button-smasher-consol-gamer whom doesn't really know how computers and softwares are works. I hope you get my point.

I have to tell you guys what really bugs me, about how computer programs in general are works nowadays, and how this very-very negatively affect, how we experience the everyday usage of a computer software.
Back in the old days if you like to use a computer (and a software) you have to learn how to use it. And it was good. It is exactly the same that why you must have to earn (by learning to drive) a driver license before you can drive a car or fly an airplane. Or just to be a doctor, first you must learn a lot. And if you mess with this things without the knowledge or proper license, you can get into really bad situations, and only you will be the responsible.
But on computers it isn't the case. It seems that almost any cases the user never responsible for anything, even if she/he who broke  something. Things how works nowadays in computers go against any common sense!
Of course, it is a good thing that everybody easily can use computers and programs, but now things are gone too far I think. Now this whole thing isn't about how we make the things easier, but how we dumb thing down so much and restrict the above-average user from changing how they experience the usage of their own computers and programs. Why we have to restrict the user as much we can and block them from truly access their computers, programs? Why the way-below-average (I like to use another word here, but I don't want to be profane) users who don't want to learn anything, are those who, at the end, decide how we supposed to use our devices and how much we can access it?
I hope one day this will change and we again can use our own computer without any restrictions so far...Thanks, if you read this "outrage".

Anyway, from the mod,modpack,subgame author side, is that this feature really this so hard to accept and use something to indicate (in some way this is already there, see 'depends.txt') what mods that really cannot be disabled or which ones can, without breaking the game?
Is this feature really that dangerous, when we already have full access to everything without any restrictions (advanced settings)? I always thought, that advanced settings are for the "geeks", not for the every-day user, so if I mess with those settings and broke something, that is only my problem. I think in many-many software I saw something a warning that if I mess with advanced settings and broke something, than I can't even get any support from the makers. I touched something that I ain't supposed to without the proper knowledge and broke it; I lost the warranty.

I really hope, someday, this feature will be implemented, regardless of the those who don't like the idea.

**Comment by rubenwardy** [comment]
Such a feature would not be in advanced settings, but would be in Configure World. The reason for this is that advanced settings is for global settings, whereas configure world allows you to enable and disable mods per world - which isn't done with a setting

**Comment by benrob0329** [comment]
>That file can be used to check if a mod required by other mods (depends on it) so we can know which mod cannot be disabled.

But that does not tell you if certain portions of the mod can be disabled, or if certain other game features need to be disabled because of this mod.

>I think we only need is that list that the subgame author provide and contains the mods that cannot be disabled.

Again, this is too blunt of a tool for game makers. There are often several aspects to a given mod and you cannot simply unload it in it's entirety without potentially causing issues for other mods of overall gameplay. We have game-specific settings for this very reason, they only need to be made per-world.

>If some is so stubborn and can't understand that; if you cannot disable a mod normally (when you already selected the subgame and select the word you want to play, then click on the configure button) but go into advanced settings and forcibly disable a must-have mod...Well, than the problem is on the user side, there must be something seriously wrong in that persons head...

My head works perfectly fine, but thank you for your concern :-)
Going into **All Settings** (not advanced, it was renamed for a reason) and changing a setting for the game is not some crazy abnormality, it's the intended way of configuring per-game and per-mod settings to customize them. **Yes**, it need to be made more accessible and _yes_, it needs to be made per-world but forcefully disabling in-game mods when we already have ways of customizing gameplay is not the way to do it.

**Comment by Rencer** [comment]
> But that does not tell you if certain portions of the mod can be disabled, or if certain other game features need to be disabled because of this mod, or if something _should_ use that mod but can work without it.

That is exactly why I keep repeating, that subgame makers must provide some-sort of a list that contains all the mods that we must not disable. The rest is up to the coders how they can implement this feature.

>...There are often several aspects to a given mod and you cannot simply unload it in it's entirety without potentially causing issues for other mods or overall gameplay. We have game-specific settings for this very reason, they only need to be made per-world.

I don't mean I want this function in somewhere in the global settings, but exactly only there in the game specific settings. Originally that is why I opened this issue, that I like to enable, or disable certain mods, when I select the already generated world, and go into the configure menu of that wold; per-world.

The rest is just about; that I exactly know that some mods never can be disabled.
I shared my thoughts about; as a simple user, how I imagine that this function can work, and what files and information I found about this subject, while I browsed through the subgames, mods folders.

I don't really have anything else to say about this matter anymore.
I miss this feature, so I always will say, that I like to see this to be in Minetest.
It's up to you guys if you implement it or not...

**Comment by ExeVirus** [comment]
So to boil down this thread, there are two opinions I see:

1. Make sub-mods (i.e. parts of mods) easily disable-able, as this is useful

2. Don't make sub-mods easily disable-able, as this will allow the everyday player to break the game easily.

----
I'm inclined to agree with 1. over 2. as Rencer has stated this would only occur for actual optional sub-mods that are tagged as such by the mod-creator, therefore eliminating the issue of 'breaking the game', but lets work through the arguments to see if there is any merit:

Now, the counter-example is that TNT mod stuff is used by other mods. I would argue this is bad planning, as the issue is right before our eyes: TNT implements stuff we want to use for other mods and that functionality should be moved to a common.lua file and not be held in the tnt.lua file. Then there would be no counter-example. _I.e. the counter example seems implemented incorrectly in the first place._

Then there is the argument that this is "too blunt of a solution". I have not read enough mods yet (maybe 30+, like techpack, terumet, mesecon, etc.) to be certain if this is true or not, but I can say with some experience that the only time this issue would crop up with mods is with dependencies outside of a mod. For example, multiple mods rely on _extraores_ on the BlockySurvival Server. If you then disabled, say mithril.lua (if that's a thing) and kept the rest of extraores, other mods that use extraores as a dependency would assume that everything is satisfied in terms of decencies only to find out that mithril, which they rely on, is missing. An error would be thrown, and the game would not run. Then someone would say "hey, where is the issue, I have already loaded extraores...". This means that while a given modpack may not rely on parts of itself, therefore allowing them to be optional, other mods might rely on the optional parts, thereby making them dependent. And there is no way to easily specify these specific dependencies without making mods really hard to update. As now every time a mod maker wnat to change the file structure or naming within a mod, it will break outside mods. Just a big headache for a medium gain.

In light of this second argument, I would say the balance tips in favor of opinion 2.  It just makes breaking mod dependencies too easy for young or inexperienced players, which is why the DB was implemented (to make things easy for this group of people). I hope this all makes sense @Rencer , as I think you are working hard to make minetest better and hate to see you get too discouraged.

**Comment by benrob0329** [comment]
>That is exactly why I keep repeating, that subgame makers must provide some-sort of a list that contains all the mods that we must not disable. The rest is up to the coders how they can implement this feature.

You completely ignored the point of the quoted statement. Yes, game makers could specify what mods are required but that doesn't help them know if certain parts of a mod are disabled. Yes, they could split mods up into multiple portions but then that can bloat things up as different disable-able parts might need common functions or different files may be run depending on whats enabled and having code to detect all of this is more complicated when you essentially deprecate the ability to detect a setting in order to know if something is enabled.

>I don't mean I want this function in somewhere in the global settings

**Exactly**, I even stated this in the quoted text.

>I shared my thoughts about; as a simple user

And I've shared my thoughts as a mod and game developer, as well as a small server operator. I agree that this needs to be fixed, just by reusing what we already have and fixing two problems with one change (per-world settings and easier game customization)

>there is no way to easily specify these specific dependencies without making mods really hard to update

Checking if a setting is enabled and erroring out isn't difficult, and good mods will keep these settings' names the same for conf compatibility.

I'd like to reiterate something that seems to be being ignored, **Mod Settings Need To Be Made Per-World Too, And They Need To Be Shown In The "Customize" Menu**. This would solve the issue from a player point-of view without any fancy sub-mods or game-breakage _and_ would solve the issue of needing custom settings files (like the Bitchange mod) for per-world settings.

Settings simply allow safer, and further customization than disabling mods entirely and moving them into per-world settings too allows players to customize other aspects of the game.
